### DISCUSSION TITLE
Freezes with forEach and fast-scroll

### DISCUSSION THREAD

**Discussion by EvsenevDev** [pointfreeco/swift-composable-architecture#1315] [Q&A]
Hi, i'm trying to use TCA for LazyVGrid, but here is issue with freezing, once you start using fast scroll or just scrolling grid fast. Once I remove send(action) it works good. I'm sure i'm not using correctly. But looks like send action have issue with a lot of reducers.

```
.combined(with: Reducer<AssetModel, GalleryPhotoAction, GalleryPhotoEnvironment> { assetModel, action, environment in
        switch action {
        case .openPhoto:
            return .none

        case .downloadImage(let size, let asset):
            return .none
        case .setImage(let image):
            assetModel.thumbnail = image
            return .none
        }
    }.forEach(
        state: \.galleryAssets,
        action: /HomeAction.gallaryPhoto(id:action:),
        environment: { GalleryPhotoEnvironment(photoLibraryService: $0.photoLibraryService) })
    )
```
    
    
View

```
struct GalleryPhotoView: View {
    let store: Store<AssetModel, GalleryPhotoAction>

    var body: some View {
        WithViewStore(store) { store in
            let cellWidthHeight = screenSize.width/CGFloat(CELLS_IN_LINE)
            let cellSize = CGSize(width: cellWidthHeight, height: cellWidthHeight)
            GeometryReader {
                if let thumbnail = store.thumbnail {
                    NavigationLink(destination: ResultView(input: .asset(store.state))) {
                        Image(uiImage: thumbnail)
                            .resizable()
                            .scaledToFill()
                    }
                } else {
                    ProgressView()
                        .frame(width: $0.size.width, height: $0.size.height, alignment: .center)
                }
            }
            .onAppear { store.send(.downloadImage(cellSize, store.asset)) }
            .aspectRatio(1, contentMode: .fit)
            .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
        }
    }
}
```

Sorry, it's duplicates https://forums.swift.org/t/tca-freezes-with-foreach-and-fast-scroll/59995

Seems like SwiftUI issue, it works good with LazyVStack, but not with LazyVGrid :(

**Comment by mbrandonw** [comment]
Interesting. Long ago we saw performance problems with lazy stacks when they were first introduced, so maybe those problems have been fixed and now the new lazy grid APIs have performance problems.

As an aside, I'm curious whether this image download logic is appropriate to run through your reducer. Is there any part of your application's logic that actually needs access to the image data? Or is the image data only being used to display in an image view?

If it's the latter, then we recommend using something like `AsyncImage` or one of the many open source URL image libraries out there. That will save you from sending actions into the system that aren't really necessary.

There's a [recent discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions/1290) thread about performance in TCA and sending many actions that you may be interested in. And you may be interested in our [article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance) discussing TCA performance.

**Reply by EvsenevDev** [reply]
Hey, thanks for answer. It makes sense to avoid sending actions, it should help me. 
Actually, yes, it's just for presenting. But it's from PhotoLibrary or iCloud (`PHImageManager`).
I'll try to find workaround for it. 
But it may be an issue in future. Maybe it will be useful to play more with `LazyVGrid` and TCA.
I've tried to use `LazyVStack` and it works good, issue reproduces just with `LazyVGrid`
